src/System.Private.CoreLib/shared/System/Globalization/CultureInfo.cs

```diff
@@ -396,9 +381,107 @@ internal static bool VerifyCultureName(CultureInfo culture, bool throwException)
+        //      o   Use WinRT to return the current user profile language
+        //      o   use current thread culture if the user already set one using CurrentCulture/CurrentUICulture
+        //      o   use thread culture if the user already set one using DefaultThreadCurrentCulture
+        //          or DefaultThreadCurrentUICulture
+        //      o   Use NLS default user culture
+        //      o   Use NLS default system culture
+        //      o   Use Invariant culture
+        //
+        public static CultureInfo CurrentCulture
+        {
+            get
+            {
+                CultureInfo ci = GetUserDefaultCultureCacheOverride();
+                if (ci != null)
+                {
+                    return ci;
+                }
+
+                if (s_currentThreadCulture != null)
+                {
+                    return s_currentThreadCulture;
+                }
+
+                ci = s_DefaultThreadCurrentCulture;
+                if (ci != null)
+                {
+                    return ci;
+                }
+
+                return s_userDefaultCulture ?? InitializeUserDefaultCulture();
+            }
+
+            set
+            {
+                if (value == null)
+                {
+                    throw new ArgumentNullException(nameof(value));
+                }
+
+                if (SetGlobalDefaultCulture(value))
+                {
+                    return;
+                }
+
+                if (s_asyncLocalCurrentCulture == null)
+                {
+                    Interlocked.CompareExchange(ref s_asyncLocalCurrentCulture, new AsyncLocal<CultureInfo>(AsyncLocalSetCurrentCulture), null);
+                }
+                s_asyncLocalCurrentCulture.Value = value;
+            }
+        }
+
+        public static CultureInfo CurrentUICulture
+        {
+            get
+            {
+                CultureInfo ci = GetUserDefaultCultureCacheOverride();
+                if (ci != null)
+                {
+                    return ci;
+                }
+
+                return GetCurrentUICultureNoAppX();
+            }
+
+            set
+            {
+                if (value == null)
+                {
+                    throw new ArgumentNullException(nameof(value));
+                }
+
+                CultureInfo.VerifyCultureName(value, true);
+
+                if (SetGlobalDefaultCulture(value))
+                {
+                    return;
+                }
+
+                if (s_asyncLocalCurrentUICulture == null)
+                {
+                    Interlocked.CompareExchange(ref s_asyncLocalCurrentUICulture, new AsyncLocal<CultureInfo>(AsyncLocalSetCurrentUICulture), null);
+                }
+
+                // this one will set s_currentThreadUICulture too
+                s_asyncLocalCurrentUICulture.Value = value;
+            }
+        }
+
         internal static CultureInfo GetCurrentUICultureNoAppX()
```

Comment:
CurrentUICulture will call GetUserDefaultCultureCacheOverride twice now. Once in the property itself, and second time here.

I think it is ok to change the ResourceManager to do what it does in CoreRT/ProjectN, and get rid of the `GetCurrentUICultureNoAppX` method (fold it into the property).

Reply:
I tried to avoid making functional changes and refactoring in the same PR. If you insist on removing `GetCurrentUICultureNoAppX` I will have to modify ResourseManager first and submit that as separate PR.

Reply:
>  modify ResourceManager first and submit that as separate PR.

Sounds good to me